doc/sphinx/techref/pyextend.rst

```diff
@@ -0,0 +1,438 @@
+:py:attr:`fontforge.font.selection` property. Many useful font-level tools will
+be selection-agnostic but some should only apply to selected glyphs.  These
+should also have enable_functions analogous to those of a glyph-level tool.
+
+Storing Per-Font and Per-Glyph Information
+------------------------------------------
+
+The FontForge SFD file format and python API includes
+:py:attr:`fontforge.font.persistent` for storing per-font information and
+:py:attr:`fontforge.glyph.persistent` for storing per-glyph information.
+These are fragile interfaces, however, and should only be used when strictly
+necessary.
+
+The FontForge program does not enforce any pattern of use on these attributes.
+We suggest the following conventions:
+
+1. If the contents are *empty*, create a dictionary and add an entry to it
+   with your plugin name as the key. Store your data in or "under" the value
+   of that entry. Then assign it to the ``persistent`` attribute.
+2. If the content is a dictionary add your entry to it if it is not already
+   present and store the result.
+3. If the content is not a dictionary warn the user with an
+   :py:meth:`fontforge.ask` dialog that allows them to opt-out of overwriting
+   the value. If they give permission you can create a dictionary as in step
+   1 and overwrite the existing data.
+
+.. _fontforge.plugin_api:
+
+Writing a FontForge Plugin
+--------------------------
+
+Even when you eventually plan on writing a FontForge Python plugin it will
+generally be easiest to start by writing an "Init Script". This is a script
+that, when placed in one of the directories listed by the
+:py:meth:`fontforge.scriptPath() <fontforge.scriptPath>` function, is
+automatically run when FontForge is started.
+
+As a simple example, here is a script that adds an "Add Midpoint Contour" entry
+to the Char View Tools menu. When any two on-curve points of a glyph are
+selected this adds a new contour containing a single point that is located
+midway between the two selected points.::
+
+    import fontforge
+
+    def getSelectedPoints(l):
+        pl = []
+        for c in l:
+            pl.extend( [ p for p in c if p.on_curve and p.selected ] )
+        return pl
+
+    def midContourEnable(u, glyph):
+        return len(getSelectedPoints(glyph.layers[glyph.activeLayer]))==2
+
+    def addMidContour(u, glyph):
+        layer_id = glyph.activeLayer
+        l = glyph.layers[layer_id]
+        pl = getSelectedPoints(l)
+        if len(pl) != 2:
+            fontforge.postError("Bad selection", "You must select "
+            "exactly two on-curve points to add a midpoint contour.")
+            return
+        nc = fontforge.contour()
+        nc.insertPoint(((pl[0].x+pl[1].x)/2, (pl[0].y+pl[1].y)/2, True,
+                fontforge.splineCorner, True))
+        l += nc
+        glyph.preserveLayerAsUndo(layer_id)
+        glyph.layers[layer_id] = l
+
+    fontforge.registerMenuItem(addMidContour, midContourEnable,
```

Comment:
I'm not sure I understand the question. Is it "What happens if you otherwise make your package discoverable but don't put key statements inside `fontforge_plugin_init`?"?

From the FontForge perspective what happens is that the plugin will look like it's failing and be reported as failing in the window even though it still registers menu items and so forth. 

It also means (and this is more important) that if another module imports the plugin it will start interacting with the FontForge interface. Someone might make a module that is both a plugin and also "offers" other useful code and the two shouldn't mix.

More generally a normal python package shouldn't have substantial "import side effects" beyond some self-configuration.

Reply:
> Is it "What happens if you otherwise make your package discoverable but don't put key statements inside fontforge_plugin_init?"?

Pretty much. Basically thinking through the security+practical implications (which basically should be you have 100% trust in said plugins, especially given how the rest of FontForge has been written/behaves) of malformed plugins

Reply:
I would say that you have to trust every *enabled* plugin. If you set the startup mode preference to "off" or "ask" FontForge will "detect" the plugin but only ever look at its metadata. (This was intentional and for security reasons.) So if want to be conservative you can only enable those plugins that you recognize. 

You could get fooled by a plugin making itself look like a different plugin but there's some protection from that because the system also tracks module name, so to actually get enabled the bad actor would have to supply a matching module in its package. That's a means of "attack" created by any script or module importing another module so there's nothing particularly special about FontForge WRT it.